src/lang/base/TypeChecker.ml

```diff
@@ -224,25 +224,36 @@ module ScillaTypechecker
         pure @@ (TypedSyntax.TApp (add_type_to_ident tf tf_rr, arg_types), (mk_qual_tp res_type, rep))
     | Message bs ->
         let%bind msg_typ = get_msgevnt_type bs in
-        let payload_type pld =
+        let payload_type fld pld =
+          let check_field_type expected =
+            match Caml.List.assoc_opt fld CU.msg_mandatory_field_types with
+            | Some fld_t when fld_t <> expected ->
+              fail1 (sprintf "Type mismatch for Message field %s" fld) (ER.get_loc rep)
+            | _ -> pure ()
+          in
           (match pld with
-           | MTag m -> pure @@ TypedSyntax.MTag m
+           | MTag m ->
+             (* If the field has a pre-determined type, it can only be string_typ. *)
```

Comment:
This is an interesting observation, because it suggests that we don't actually need the `MTag` constructor. We can just use `MLit` and check that the literal is a string.

I think `MTag` comes from a time when the `_tag` entry could have a non-string value (like `{ _tag : Main ; ...}`), but we don't really use that form anymore.

It is convenient to have the `MTag` form if we ever want to check that a recipient contract has a transition with the correct name, but that seems like a higher-level check, and since a higher-level language could just compile its tags down to an `MLit` entry, it seems unnecessary at Scilla level.

Reply:
Noted through #434.